File: 8/8.py
from tqdm import tqdm

'''
Given a list of instructions, it will return a dictionary that looks like this that represents the graph
dictA = {'AAA': ['BBB', 'BBB'], 'BBB': ['AAA', 'ZZZ'], 'ZZZ': ['ZZZ', 'ZZZ']}

'''
def parseInstructions(instructions):
    nodes = {}
    for line in instructions:
        [source, dest] = line.split(' = ')
        [left, right] = dest.split(', ')
        nodes[source] = [left[1:], right[:len(right.rstrip()) - 1]]
    return nodes




dictA = {'A': ['B', 'B'], 'B': ['A', 'ZZZ'], 'Z': ['Z', 'Z']}
dictB = {'A': ['B', 'C'], 'B': ['D', 'E'], 'C': ['Z', 'G'], 'D': ['D', 'D'], 'E': ['E', 'E'], 'G': ['G', 'G'], 'Z': ['Z', 'Z']}

def to_dot(graph):
    dot = "digraph {\n"
    for node, edges in graph.items():
        for edge in edges:
            dot += f'    "{node}" -> "{edge}";\n'
    dot += "}"
    return dot


ex = '''
LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
'''
def main():
    inputFile = list(open('input.txt'))
    instructions = inputFile[0].rstrip()
    graph = parseInstructions(inputFile[2:])
    frontier = {x for x in graph.keys() if x[-1] == 'A'}
    frontier_last_chars = {x[-1] for x in frontier}
    # Do i need to check cycles?
    idx = 0
    iters = 100000000000 
    with tqdm(total=iters) as pbar:
        while idx < iters:
            if all(x == 'Z' for x in frontier_last_chars):
                return idx
            instruction = instructions[idx % len(instructions)]
            selection = 0 # assume 'L'
            if instruction == 'R':
                selection = 1
            frontier = {graph[x][selection] for x in frontier}
            frontier_last_chars = {x[-1] for x in frontier}
            idx += 1
            pbar.update(1)
        return idx, set(frontier)
print(main())

File: 8/8.js
const fs = require('fs');

function parseInstructions(instructions) {
    let nodes = {};
    for (let line of instructions) {
        let [source, dest] = line.split(' = ');
        let [left, right] = dest.split(', ');
        nodes[source] = [left.slice(1), right.slice(0, right.length - 1)];
    }
    return nodes;
}

async function finder(node, graph, instructions) {
  let idx = 0;
  let iters = 1000000000;
  idx = 0
  while (idx < iters) {
    if (node[node.length - 1] === 'Z') {
        return idx;
    }
    switch (instructions[idx % instructions.length]) {
        case 'L':
            node = graph[node][0];
            break;
        case 'R':
            node = graph[node][1];
            break;
    }
    idx += 1;
  }
  return idx
}

async function main() {
    let inputFile = fs.readFileSync('input.txt', 'utf-8').split('\n');
    let instructions = inputFile[0].trim();
    let graph = parseInstructions(inputFile.slice(2));
    let frontier = Object.keys(graph).filter(x => x.slice(-1) === 'A');

    // Got this from https://stackoverflow.com/questions/31302054/how-to-find-the-least-common-multiple-of-a-range-of-numbers
    const gcd = (a, b) => b == 0 ? a : gcd (b, a % b)
    const lcm = (a, b) =>  a / gcd (a, b) * b
    const lcmAll = (ns) => ns .reduce (lcm, 1)

    const pathLengths = await Promise.all(frontier.map(async (node) => {
      return await finder(node, graph, instructions)
    }))
    return lcmAll(pathLengths)
    // Because the instructions are technically infinite repeating sequences, we can just find the LCM of the path lengths
    // to find the first time they all repeat. It is much much faster to do this numerically than it is to try traversing each one. 
    // I ended up writing the solution to part two in js, because of the perceived benefits in parallelism. However I learned that
    // js is single threaded, so tbd how much it actually ended up helping me :/ 
}

main().then(console.log);